### PR TITLE
BUG : fix non-uniform grids in pcolorfast

### LINKED ISSUE #4227: pcolorfast fails in master when image is not uniform
Hi matplotlib developers,

When using master with python 2.7 on Linux, I discovered that `pcolorfast` fails if the image grid is not uniformly spaced.  Here is a minimal working example:

```python
import numpy as np
import matplotlib.pyplot as plt

Z = np.arange(6).reshape((3, 2))
X = np.array([0,1,2,10])  # <-- nonuniform array
Y = np.array([0,1,2])

plt.figure()
ax = plt.subplot(111)
ax.pcolorfast(X,Y,Z.T)  # <-- fails
# ax.pcolormesh(X,Y,Z.T)  # <-- works

plt.show()
```

This works fine in version 1.4.3, but fails in master with the following traceback:
```python
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/matplotlib/backends/backend_qt5.py", line 341, in resizeEvent
    self.draw()
  File "/usr/lib/python2.7/site-packages/matplotlib/backends/backend_qt5agg.py", line 143, in draw
    FigureCanvasAgg.draw(self)
  File "/usr/lib/python2.7/site-packages/matplotlib/backends/backend_agg.py", line 475, in draw
    self.figure.draw(self.renderer)
  File "/usr/lib/python2.7/site-packages/matplotlib/artist.py", line 60, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/matplotlib/figure.py", line 1094, in draw
    func(*args)
  File "/usr/lib/python2.7/site-packages/matplotlib/artist.py", line 60, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/matplotlib/axes/_base.py", line 2096, in draw
    a.draw(renderer)
  File "/usr/lib/python2.7/site-packages/matplotlib/artist.py", line 60, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/matplotlib/image.py", line 881, in draw
    im = self.make_image(renderer.get_image_magnification())
  File "/usr/lib/python2.7/site-packages/matplotlib/image.py", line 869, in make_image
    bg)
TypeError: integer argument expected, got float
```

The same failure occurs with all of the backends that I have tried; which are specificially the `wxagg`, `gtkagg`, `tkagg`, and `pdf` backends.

Thanks for the help.

### PR DESCRIPTION
**Pull request by tacaswell**
Closes #4227 

attn @mdboom @efiring  This fixes the problem, but I have no idea why (I just looked at the c++ header, it wanted ints so I made sure it get ints).

### CODE REVIEW

**Comment by tacaswell** [comment]
We might also want to back-port this to color-overhaul.

File: lib/matplotlib/image.py [hunk 1]
```diff
@@ -852,8 +852,8 @@ def make_image(self, magnification=1.0):
         l, b, r, t = self.axes.bbox.extents
         width = (round(r) + 0.5) - (round(l) - 0.5)
         height = (round(t) + 0.5) - (round(b) - 0.5)
-        width = width * magnification
-        height = height * magnification
+        width = int(width * magnification)
+        height = int(height * magnification)
```
Comment (efiring) [inline lib/matplotlib/image.py]: Looks good to me.  The only question it raises in my mind is whether after multiplication by magnification, a rounding operation would be better than truncation.

**Comment by efiring** [comment]
There is no point in backporting because the CXX version was handling the conversion to int.  That also tells me there is no need to add a rounding step; what you have is fine.
Another way to fix the problem would be to put the conversion to int back in the C++; that makes the _pcolor2 function friendlier, but given that it is private, I have no objection to letting it be fussy about inputs.  It's simpler.

**Comment by tacaswell** [comment]
@efiring rebased to fix conflict in tests.

Re-reading this I read your suggestion about rounding, implemented it and then re-read your comment about not implementing it.

I think the rounding is ok as this is determining how many pixels in the agg layer that the image will take up.  Assuming travis passes, I think we should leave it as-is.

**Comment by tacaswell** [comment]
well, of course travis is going to pass, we have no test coverage on this...

**Comment by efiring** [comment]
Needs a rebase?  Otherwise, it looks fine.

**Comment by tacaswell** [comment]
@efiring rebased.

**Pull request merged**